compiler/sem/sanitizer.nim

```diff
@@ -0,0 +1,976 @@
+    result = newTreeI(nkCast, n.info): [parseTypeExpr(c, n[0]), expr(c, n[1])]
+  of nkStaticExpr:
+    # skip
+    checkSonsLen(n, 1)
+    result = expr(c, n[0])
+  of nkAddr:
+    checkSonsLen(n, 1)
+    result = newTreeI(nkAddr, n.info): expr(c, n[0])
+  of nkHiddenAddr, nkHiddenDeref:
+    checkSonsLen(n, 1)
+    result = expr(c, n[0])
+  of nkObjDownConv, nkObjUpConv:
+    # XXX: these don't exist in typed AST prior to ``transf``. How should they
+    #      be treated?
+    invalid()
+  of nkChckRange, nkChckRangeF, nkChckRange64, nkStringToCString, nkCStringToString:
+    # XXX: neither do these
+    invalid()
+  of nkAsgn, nkFastAsgn:
+    checkSonsLen(n, 2)
+    result = newTreeI(n.kind, n.info): [expr(c, n[0]), expr(c, n[1])]
+  of nkGenericParams, nkFormalParams:
+    unreachable("missing")
+  of nkOfInherit:
+    unreachable("missing")
+  of nkImportAs:
+    # turn it back into an infix
+    checkSonsLen(n, 2)
+    result = newTreeI(nkInfix, n.info): [ident(c, "as"), expr(c, n[0]), expr(c, n[1])]
+  of nkProcDef, nkFuncDef, nkMethodDef, nkConverterDef, nkMacroDef, nkTemplateDef, nkIteratorDef:
+    # TODO: revisit the translation here
+    if n.len-1 notin bodyPos..resultPos:
+      return invalidAstLen(c, n, bodyPos)
+
+    result = prepareFrom(n)
+    result[namePos] = emptyOr(c, n[namePos], parseDef)
+    result[patternPos] = requireEmpty(c, n[patternPos]) # TODO: missing
+    result[genericParamsPos] = requireEmpty(c, n[genericParamsPos]) # TODO: missing
+
+    case n[paramsPos].kind
+    of nkEmpty:
+      result[paramsPos] = empty(n[paramsPos].info)
+    of nkFormalParams:
+      result[paramsPos] = guardMinLen(n[paramsPos], 1):
+        var r = prepareFrom(n[paramsPos])
+        r[0] =
+          case n[paramsPos][0].kind
+          of nkEmpty: empty(n[paramsPos][0].info)
+          else: typeExpr(c, n[paramsPos][0])
+        for i, it in sliceIt(n[paramsPos], 1..^1):
+          r[i] = identDefs(c, it)
+        r
+    else:
+      result[paramsPos] = invalidAst(c, n[paramsPos])
+
+    result[pragmasPos] =
+      case n[pragmasPos].kind
+      of nkEmpty:  empty(n[pragmasPos].info)
+      of nkPragma: parsePragma(c, n[pragmasPos])
+      else:        invalidAst(c, n[pragmasPos])
+
+    # ignore the contents of the 'misc' slot
+    result[miscPos] = empty(n.info)
+    result[bodyPos] = stmt(c, n[bodyPos])
+    if n.len > resultPos:
+      result[resultPos] = empty(n.info)
+
+  of nkOfBranch, nkElifBranch, nkExceptBranch, nkElse, nkElifExpr, nkElseExpr:
+    invalid()
+  of nkAsmStmt:
+    checkSonsLen(n, 2)
+    result = newTreeI(nkAsmStmt, n.info)
+    result[0] = emptyOr(c, n, parsePragma)
+    result[1] = parseStrLit(c, n[1])
+  of nkPragma:
+    checkMinSonsLen(n, 1)
+    for i in 0..<n.len:
+      result[i] = parseColonEqExpr(c, n[i])
+  of nkPragmaBlock:
+    unreachable("missing")
+  of nkForStmt:
+    checkMinSonsLen(n, 3)
+    result = prepareFrom(n)
+    # note: this allows for code that the parser doesn't
```

Comment:
Hmm, the syntax ast and the untyped ast spec diverging is concerning. 🤔

Reply:
With how it's currently implemented, the sanitizer allows the following code:
```nim
for (a, b), (c, d), (e, f) in items: ...
```
while the parser doesn't. For what it's worth, with the recent `semForVar` refactoring, this would also be correctly analysed by sem.